Propose more sass:math module functions (#2779)
sass/language: proposal/more-math-functions.md

    `$exponent > 0`, return `Infinity` as a unitless number.

  * If `$base == -0` and `$exponent < 0`, or if `$base == -Infinity` and
    `$exponent > 0`:
    * If `$exponent` is an odd integer, return `-Infinity` as a unitless number.
    * Return `Infinity` as a unitless number.

  * If `$base == 0` and `$exponent > 0`, or if `$base == Infinity` and
    `$exponent < 0`, return `0` as a unitless number.

  * If `$base == -0` and `$exponent > 0`, or if `$base == -Infinity` and
    `$exponent < 0`:
    * If `$exponent` is an odd integer, return `-0` as a unitless number.
    * Return `0` as a unitless number.

  * Return `$base` raised to the power of `$exponent`, as a unitless number.

#### `sqrt()`

```
sqrt($number)
```

* If `$number` has units, throw an error.
* If `$number < 0`, return `NaN` as a unitless number.
* If `$number == -0`, return `-0` as a unitless number.
* If `$number == Infinity`, return `Infinity` as a unitless number.
* Return the square root of `$number`, as a unitless number.

### Trigonometry

#### `cos()`

```
cos($number)
```

* If `$number` has units but is not an [angle][], throw an error.

> * If `$number` is unitless, treat it as though its unit were rad.

Reviewer: Style nit:

```suggestion
* If `$number` is unitless, treat it as though its unit were `rad`.
```

Author: Fixed.

## Table of Contents
* [Variables](#variables)
  * [`$e`](#e)
  * [`$pi`](#pi)
* [Functions](#functions)
  * [`clamp()`](#clamp)
  * [`hypot()`](#hypot)
  * [Exponentiation](#exponentiation)
    * [`log()`](#log)
    * [`pow()`](#pow)
    * [`sqrt()`](#sqrt)
  * [Trigonometry](#trigonometry)
    * [`cos()`](#cos)
    * [`sin()`](#sin)
    * [`tan()`](#tan)
    * [`acos()`](#acos)
    * [`asin()`](#asin)
    * [`atan()`](#atan)
    * [`atan2()`](#atan2)
      * [Edge cases](#edge-cases)

## Variables

### `$e`

Equal to the value of the mathematical constant `e` with a precision of 10
digits: `2.718281828`.

### `$pi`

Equal to the value of the mathematical constant `pi` with a precision of 10
digits: `3.141592654`.

## Functions

### `clamp()`

```
clamp($min, $number, $max)
```

* Throw an error if the units of `$min`, `$number`, and `$max` are not
  [compatible][] with each other.
* Return `$min` if `$min >= $max`
* Return `$min` if `$number <= $min`.
* Return `$max` if `$number >= $max`.
* Return `$number`.

[compatible]: ../spec/built_in_modules/math.md#compatible

### `hypot()`

```
hypot($numbers...)
```

> * Throw an error if any argument is not a [length][].

Reviewer: If we don't allow unitless numbers here, I think that makes it the *only* Sass function that requires units. That's probably more confusing of an inconsistency than we'd have if we were more lenient than plain CSS and allowed unitless numbers here.

We could even go so far as to allow any compatible units here, since the output unit is always well-defined. I'm tempted to do so, because (unlike CSS) Sass allows custom units for which vector-length calculations could make sense.

Author: Yeah, makes sense to me. I'll have `hypot()` accept any arguments that are all compatible with each other, and return a value that takes its unit from the leftmost argument.

* Return `Infinity` if any argument is `Infinity`.

> * Return the square root of the sum of the squares of each argument.

Reviewer: All of these functions should define the units the return value uses, even if it's just "as a unitless number".

Author: Got it, fixed.

[length]: https://drafts.csswg.org/css-values-4/#lengths

### Exponentiation

#### `log()`

```
log($number, $base: null)
```

* Throw an error if `$number` has units or `$number < 0`.

> * If `$base == null`:

Reviewer: Style nit for consistency with existing specs:

```suggestion
* If `$base` is null:
```

Author: Thanks, fixed.

  * Return `-Infinity` if `$number == 0`.
  * Return `Infinity` if `$number == Infinity`.
  * Return the [natural log][] of `$number`.

> * Otherwise:

Reviewer: Style nit:

```md
* Otherwise, if `$base < 0` or `$base == 0` or `$base == 1`, throw an error.
* Otherwise, return the natural log of `$number` divided by the natural log of `$base`.
```

Writing "Otherwise" for each branch of the conditional helps draw the reader's attention to the fact that there are three possible cases here.

Author: Good point. Fixed.

  * Throw an error if `$base < 0` or `$base == 0` or `$base == 1`.
  * Return the natural log of `$number` divided by the natural log of `$base`.

[natural log]: https://en.wikipedia.org/wiki/Natural_logarithm

#### `pow()`

```
pow($base, $exponent)
```

* Throw an error if `$base` or `$exponent` has units.

* If `$exponent == 0`:
  * Return `1`.

* Otherwise, if `$exponent == Infinity`:

>   * Return `NaN` if the `absolute value of $base == 1`.

Reviewer: "Absolute value of" isn't really syntax, so it reads strange to have it code-highlighted... I'd probably either write this in prose:

```suggestion
  * Return `NaN` if the absolute value of `$base` equals `1`.
```

or avoid mentioning absolute value altogether:

```suggestion
  * Return `NaN` if `$base == 1` or `$base == -1`.
```

Author: Got it. I'll go with the latter.

    or if the `absolute value of $base < 1` and `$exponent < 0`.
  * Return `0`.

* Otherwise:
  * Return `NaN` if `$base < 0` and `$exponent` is not an integer.

  * Return `Infinity` if `$base == 0` and `$exponent < 0`, or if
    `$base == Infinity` and `$exponent > 0`.
  * If `$base == -0` and `$exponent < 0`, or if `$base == -Infinity` and
    `$exponent > 0`:
    * Return `-Infinity` if `$exponent` is an odd integer.
    * Return `Infinity`.

  * Return `0` if `$base == 0` and `$exponent > 0`, or if `$base == Infinity`
    and `$exponent < 0`.
  * If `$base == -0` and `$exponent > 0`, or if `$base == -Infinity` and
    `$exponent < 0`:
    * Return `-0` if `$exponent` is an odd integer.
    * Return `0`.

  * Return `$base` raised to the power of `$exponent`.

#### `sqrt()`

```
sqrt($number)
```

* Throw an error if `$number` has units.
* Return `NaN` if `$number < 0`.
* Return `-0` if `$number == -0`.
* Return `Infinity` if `$number == Infinity`.
* Return the square root of `$number`.

### Trigonometry

#### `cos()`

```
cos($number)
```

* Throw an error if `$number` has units but is not an [angle][]. If `$number` is

>   unitless, it is assumed to be in radians.

Reviewer: ```suggestion
  unitless, it is treated as though its unit were `rad`.
```

Author: Fixed.

* Return `NaN` if `$number == Infinity`.
* Return the [cosine][] of `$number`, which is a unitless number between `-1`

>   and `1`, exclusive.

Reviewer: Why exclusive? Isn't `cos(pi)` -1?

I don't think you actually need to specify the range here, since it's contained in the definition of the trig functions. If you want to for clarity, put it in a block quote so it's clear that it's non-normative.

Author: Got it. I'll just lean on the trig definition

[angle]: https://drafts.csswg.org/css-values-4/#angles
[cosine]: https://en.wikipedia.org/wiki/Trigonometric_functions#Right-angled_triangle_definitions

#### `sin()`

```
sin($number)
```

* Throw an error if `$number` has units but is not an angle. If `$number` is
  unitless, it is assumed to be in radians.
* Return `NaN` if `$number == Infinity`.
* Return `-0` if `$number == -0`.
* Return the [sine][] of `$number`, which is a unitless number between `-1` and
  `1`, exclusive.

[sine]: https://en.wikipedia.org/wiki/Trigonometric_functions#Right-angled_triangle_definitions

#### `tan()`

```
tan($number)
```

* Throw an error if `$number` has units but is not an angle. If `$number` is
  unitless, it is assumed to be in radians.
* Return `NaN` if `$number == Infinity`.
* Return `-0` if `$number == -0`.

> * Return `Infinity` if `$number == 90 +/- (360 * n)`, where `n` is any

Reviewer: ```suggestion
* Return `Infinity` if `$number % 360deg == 90deg`.
```

It's important to specify units here, since the values 90 and 360 don't mean anything on their own if `$number` is in radians.

Author: Fixed.

* Return `NaN` if the `absolute value of $number > 1`.
* Return `0` if `$number == 1`.
* Return the [arccosine] of `$number`, in radians. It is between `0` and
  `pi rad`, inclusive.

[arccosine]: https://en.wikipedia.org/wiki/Inverse_trigonometric_functions#Basic_properties

#### `asin()`

```
asin($number)
```

* Throw an error if `$number` has units.
* Return `NaN` if `absolute value of $number > 1`.
* Return `-0` if `$number == -0`.
* Return the [arcsine] of `$number`, in rad. It is between `-pi/2 rad` and
  `pi/2 rad`, inclusive.

[arcsine]: https://en.wikipedia.org/wiki/Inverse_trigonometric_functions#Basic_properties

#### `atan()`

```
atan($number)
```

* Throw an error if `$number` has units.
* Return `-0` if `$number == -0`.
* Return `-90` if `$number == -Infinity`.
* Return `90` if `$number == Infinity`.
* Return the [arctangent] of `$number`, in rad. It is between `-pi/2 rad` and
  `pi/2 rad`, inclusive.

[arctangent]: https://en.wikipedia.org/wiki/Inverse_trigonometric_functions#Basic_properties

#### `atan2()`

> `atan2($y, $x)` is distinct from `atan($y / $x)` because it preserves the
> quadrant of the point in question. For example, `atan2(1, -1)` corresponds to

> > the point `(-1, 1)` and returns `(0.75 * pi) rad`. In contrast, `atan(1 / -1)`

Reviewer: Consider making these radian examples valid SassScript:

```suggestion
> the point `(-1, 1)` and returns `0.75rad * $pi`. In contrast, `atan(1 / -1)`
```

Author: Makes sense, thanks.

> and `atan(-1 / 1)` resolve first to `atan(-1)`, so both return (`-0.25 * pi)

> > rad`.

Reviewer: This is a really good explanation :+1:.

Author: Thanks 🙂

```
atan2($y, $x)
```

> * `$y` and `$x` must either have the same unit or be unitless. If not, throw an

Reviewer: Style nit: For consistency, phrase this as either "If X, throw an error" or "Throw an error if X".

Author: Will go with `If X, throw an error`, here and elsewhere.

> * `$y` and `$x` must either have the same unit or be unitless. If not, throw an

Reviewer: Shouldn't this just check if `$y` and `$x` have compatible units?

Author: Yes, good catch!

> # Math Module

Reviewer: Can you format the intro more like [other proposals](https://github.com/sass/sass/blob/master/proposal/import-configuration.md), with a short proposal name, draft number, and link to the GitHub issue?

Author: Done.

This built-in module is available from the URL `sass:math`.

## Table of Contents
* [Variables](#variables)
  * [`e`](#e)
  * [`pi`](#pi)
* [Functions](#functions)
  * [Approximation](#approximation)
    * [`ceil()`](#ceil)
    * [`floor()`](#floor)
    * [`round()`](#round)
  * [Comparison](#comparison)
    * [`clamp()`](#clamp)
    * [`max()`](#max)
    * [`min()`](#min)
    * [`percentage()`](#percentage)
  * [Division](#division)
    * [`div()`](#div)
    * [`mod()`](#mod)
  * [Exponentiation](#exponentiation)
    * [`log()`](#log)
    * [`pow()`](#pow)
    * [`sqrt()`](#sqrt)
  * [Trigonometry](#trigonometry)
    * [Procedures](#procedures)
      * [Verify Inputs](#verify-inputs)
    * [`cos()`](#cos)
    * [`sin()`](#sin)
    * [`tan()`](#tan)
    * [`acos()`](#acos)
    * [`asin()`](#asin)
    * [`atan()`](#atan)
    * [`atan2()`](#atan2)
      * [Edge cases](#edge-cases)
  * [Units](#units)
    * [`compatible()`](#compatible)
    * [`is-unitless()`](#is-unitless)
    * [`unit()`](#unit)
  * [Util](#util)
    * [`abs()`](#abs)
    * [`hypot()`](#hypot)
    * [`random()`](#random)

## Variables

> ### `e`

Reviewer: Document the values these variables should have.

Author: Done.

```

> $e

Reviewer: I don't think you need to include signatures for these, since they're just variable names. I'd probably include the `$` sign in the header title though.

Author: Done.

```

### `pi`

```
$pi
```

## Functions

### Approximation

> #### `ceil()`

Reviewer: Proposals are usually written as modifications to the existing spec, so you don't need to include functions that already exist—you can just write something like "This proposal adds the following members to the `sass:math` module" at the top.

Author: Done.

```
ceil($number)
```
This function is also available as a global function named `ceil()`.

#### `floor()`

```
floor($number)
```

This function is also available as a global function named `floor()`.

#### `round()`

```
round($number)
```

This function is also available as a global function named `round()`.

### Comparison

#### `clamp()`

```

> clamp($min, $number, $max)

Reviewer: How does this handle the numbers having incompatible units?

Author: This should throw an error

```

> * Return `$min` if `$min` >= `$max`

Reviewer: Style nit:

```suggestion
* Return `$min` if `$min >= $max`
```

All programming or mathematical syntax should be written in code font.

Author: Agreed, fixed throughout the document

* Return `$min` if `$number` <= `$min`.
* Return `$max` if `$number` => `$max`.
* Return `$number`.

#### `max()`

```
max($numbers...)
```

This function is also available as a global function named `max()`.

#### `min()`

```
min($numbers...)
```

This function is also available as a global function named `min()`.

#### `percentage()`

```
percentage($number)
```

This function is also available as a global function named `percentage()`.

### Division

#### `div()`

```
div($dividend, $divisor)
```

This function is also available as a global function named `divide()`.

* Throw error if the units of `$dividend` and `$divisor` are not
  [compatible](#compatible).

> * Return `$dividend` / `$divisor`.

Reviewer: I think you can leave this out, since it's already covered by [the slash-separator proposal](https://github.com/sass/sass/blob/master/accepted/slash-separator.md#dividemathdiv-function).

Author: Removed.

#### `mod()`

```
mod($dividend, $modulus)
```

> * Throw error if the units of `$dividend` and `$modulus` are not

Reviewer: Style nit:

```suggestion
* Throw an error if the units of `$dividend` and `$modulus` are not
```

I find full sentences help specs read more clearly, especially for people unfamiliar with their formality.

Author: Got it, fixed.

  [compatible](#compatible).

> * Return `$dividend` % `$modulus`.

Reviewer: Since `%` isn't (currently) a Sass operator, I think this operation needs to be defined in more depth, especially with respect to what units the returned number has.

Author: Discussed offline: since `%` exists as a binary operation, and because it doesn't seem likely that CSS will introduce new semantics for `%`, we can just remove this `modulo` proposal.

### Exponentiation

#### `log()`

```
log($number)
```
* Throw error if `$number` has units or `$number` < 0.
* Return -Infinity if `$number` is 0.
* Return Infinity if `$number` is infinite.

> * Return the natural log of `$number`.

Reviewer: You don't need to precisely define what the natural logarithm means, but it's probably worth linking to the Wikipedia article.

Author: Added link here and wherever needed

```

> log($number, $base)

Reviewer: It's usually a good idea to avoid defining overloads unless it's really necessary, since they aren't representable in pure Sass and thus can be confusing to read. For example, you could define this function as `log($exponent, $base: null)` and have it return the natural logarithm if `$base` is null. That also has the benefit that a user can explicitly pass a null base if they want to trigger the default behavior.

Author: Ah, good point. Fixed.

```

* Throw error if `$base` < 0 or `$base` is 0 or `$base` is 1.

> * Return [log](#log)(`$number`) / [log](#log)(`$base`).

Reviewer: I think it would be clearer to just write "Return the natural log of `$number` divided by the natural log of `$base`" rather than referring to other Sass functions. Similarly below.

Author: Done.

#### `pow()`

```
pow($base, $exponent)
```

* Throw error if `$base` or `$exponent` has units.

* If `$exponent` is 0:
  * Return 1.

> * Else if `$exponent` is infinite:

Reviewer: ```suggestion
* Otherwise, if `$exponent` is infinite:
```

This just matches the existing spec language.

Author: Done.

> * Else if `$exponent` is infinite:

Reviewer: I'd explicitly compare `$exponent` to `Infinity` and `-Infinity` here, just to avoid the term "infinite". A reader will *probably* know what it's referring to, but they could wrongfully assume that it's a more complex term of art.

Author: Fixed here and throughout

  * Return NaN if [abs](#abs)(`$base`) is 1.
  * Return Infinity if [abs](#abs)(`$base`) > 1 and `$exponent` > 0, or if
    [abs](#abs)(`$base`) < 1 and `$exponent` < 0.
  * Return 0.

* Else if `$exponent` is finite:
  * Return NaN if `$base` < 0 and `$exponent` is not an integer.

  * Return Infinity if `$base` is 0 and `$exponent` < 0, or if `$base` is
    Infinity and `$exponent` > 0.
  * If `$base` is -0 and `$exponent` < 0, or if `$base` is -Infinity and
    `$exponent` > 0:
    * Return -Infinity if `$exponent` is an odd integer.
    * Return Infinity.

  * Return 0 if `$base` is 0 and `$exponent` > 0, or if `$base` is Infinity and
    `$exponent` < 0.
  * If `$base` is -0 and `$exponent` > 0, or if `$base` is -Infinity and
    `$exponent` < 0:
    * Return -0 if `$exponent` is an odd integer.
    * Return 0.

  * Return `$base` raised to the power of `$exponent`.

#### `sqrt()`

```
sqrt($number)
```

* Throw error if `$number` has units.
* Return NaN if `$number` < 0.
* Return -0 if `$number` is -0.
* Return Infinity if `$number` is infinite.
* Return the square root of `$number`.

### Trigonometry

> #### Procedures

Reviewer: Procedures should go at the top of the proposal, as a sibling section to Functions.

Author: Done.

> ##### Verify Inputs

Reviewer: This should follow the same format as [other spec procedures](https://github.com/sass/sass/pull/2777/files?short_path=ba3ae5f#diff-ba3ae5f01809e2c7eea84ff84e951e24#user-content-determining-a-use-rules-namespace), with a prose line indicating which inputs are taken and a bulleted list of steps to take.

If the procedure differs from function to function, though, it would probably be clearer to just include each function's behavior inline in the function definition. A bit of repetition is usually fine in prose.

Author: Like you said, this is much clearer when each function has its own input verification instead of trying to DRY it up too much. I removed this section altogether.

`cos($number), sin($number), and tan($number)` throw error if `$number` has

> units but is not a rotation value. If `$number` is unitless, it is assumed to be

Reviewer: Define "rotation value" (linking to a CSS spec would be sufficient).

Author: Done. Linked to the precise CSS definition `angle`

in radians.

`acos($number), asin($number), and atan($number)` throw error if `$number` has
units.

Both inputs to `atan2($y, $x)` must have the same unit or be unitless. If not,
throw error.

#### `cos()`

```
cos($number)
```

* [Verify inputs](#verify-inputs).
* Return NaN if `$number` is infinite.

> * Return cos(`$number`).

Reviewer: These trigonometric functions should be more explicitly defined—provide links to definitions of what the functions do, as well as explanations of how to handle different units and what unit should be returned (probably unitless?).

Author: Done.

#### `sin()`

```
sin($number)
```

* [Verify inputs](#verify-inputs).
* Return NaN if `$number` is infinite.
* Return -0 if `$number` is -0.
* Return sin(`$number`).

#### `tan()`

```
tan($number)
```

* [Verify inputs](#verify-inputs).
* Return NaN if `$number` is infinite.
* Return -0 if `$number` is -0.

> * If `$number` is an asymptote value:

Reviewer: Define "asymptote value"

Author: I realized the word "asymptote" is kind of a red herring, I just removed this line to let the logic in the next two lines speak for themselves.

  * Return Infinity if `$number` is 90deg +/- 360*n* deg, where *n* is any integer.
  * Return -Infinity if `$number` is -90deg +/- 360*n* deg, where *n* is any
    integer.
* Return tan(`$number`).

#### `acos()`

```
acos($number)
```

* [Verify inputs](#verify-inputs).
* Return NaN if [abs](#abs)(`$number`) > 1.
* Return 0 if `$number` is 1.
* Return acos(`$number`).

#### `asin()`

```
asin($number)
```

* [Verify inputs](#verify-inputs).
* Return NaN if [abs](#abs)(`$number`) > 1.
* Return -0 if `$number` is -0.
* Return asin(`$number`).

#### `atan()`

```
atan($number)
```

* [Verify inputs](#verify-inputs).
* Return -0 if `$number` is -0.
* Return -90deg if `$number` is -Infinity.
* Return 90deg if `$number` is Infinity.
* Return atan(`$number`).

#### `atan2()`

`atan2($y, $x)` is distinct from `atan($y / $x)` because it preserves the
quadrant of the point in question. For example, `atan2(1, -1)` corresponds to
the point (-1, 1), which returns 135deg. In contrast, `atan(1 / -1)` and

> `atan(-1 / 1)` resolve first to `atan(-1)`, so both return -45deg.

Reviewer: As a convention, we write non-normative text in block quotes:

```markdown
> `atan2($y, $x)` is distinct from `atan($y / $x)` because it preserves the
> quadrant of the point in question. For example, `atan2(1, -1)` corresponds to
> the point (-1, 1), which returns 135deg. In contrast, `atan(1 / -1)` and
> `atan(-1 / 1)` resolve first to `atan(-1)`, so both return -45deg.
```

Author: Good to know, thanks

```
atan2($y, $x)
```

* [Verify inputs](#verify-inputs).
* If the inputs match one of the following edge cases, return the value for that

>   edge case. Otherwise, return normal atan2($y, $x).

Reviewer: Is this supposed to be "normal atan"?

Author: I meant "normal" as any non-edge-case input, but realize that's pretty unclear. Changed this to specify it as a "2-argument arctangent", and linked to the wiki.

##### Edge cases

> <table>

Reviewer: Can you use a [markdown table](https://github.github.com/gfm/#tables-extension-) for this?

Author: I would've liked to, but markdown tables don't allow cells to span multiple rows/cols.

      <td>-180deg</td>
      <td>-180deg</td>
      <td>-0deg</td>
      <td>-0deg</td>
      <td>-0deg</td>
    </tr>
    <tr>
      <th>0</th>
      <td>180deg</td>
      <td>180deg</td>
      <td>180deg</td>
      <td>0deg</td>
      <td>0deg</td>
      <td>0deg</td>
    </tr>
    <tr>
      <th>finite</th>
      <td>180deg</td>
      <td>(normal)</td>
      <td>90deg</td>
      <td>90deg</td>
      <td>(normal)</td>
      <td>0deg</td>
    </tr>
    <tr>
      <th>Infinity</th>
      <td>135deg</td>
      <td>90deg</td>
      <td>90deg</td>
      <td>90deg</td>
      <td>90deg</td>
      <td>45deg</td>
    </tr>
  </tbody>
</table>

### Units

#### `compatible()`

```
compatible($number1, $number2)
```

This function is also available as a global function named `comparable()`.

#### `is-unitless()`

```
is-unitless($number)
```

This function is also available as a global function named `unitless()`.

#### `unit()`

```
unit($number)
```

This function is also available as a global function named `unit()`.

### Util

#### `abs()`

```
abs($number)
```

This function is also available as a global function named `abs()`.

#### `hypot()`

```
hypot($numbers...)
```

> * Throw error if any argument is not a length value.

Reviewer: Define "length value". Does this allow unitless numbers?

Author: Done. Linked to the precise CSS definition `length`

clamp($min, $number, $max)
```

* If the units of `$min`, `$number`, and `$max` are not [compatible][] with each
  other, throw an error.
* If `$min >= $max`, return `$min`.
* If `$number <= $min`, return `$min`.
* If `$number >= $max`, return `$max`.
* Return `$number`.

[compatible]: ../spec/built_in_modules/math.md#compatible

### `hypot()`

```
hypot($arguments...)
```

* If all arguments are not compatible with each other, throw an error.
* Let the unit of the return value be the same as that of the leftmost argument
  that has a unit. If all arguments are unitless, let the return value be

>   unitless.

Reviewer: What do you think about forbidding mixing unitless numbers with numbers that have units? The lack of transitivity there is weird, and while it's probably too late to change it for `+` and `-`, we could definitely change it here.

Author: The only reason I considered allowing unitless numbers was to model the behavior after `+`/`-` 🙂 But if the lack of transitivity is undesired, I'm good with forbidding mixing unitless/units.

  * If any argument is `Infinity`, return `Infinity`.

>   * Return the square root of the sum of the squares of each argument.

Reviewer: Did you intend to indent these?

Author: I wanted to group these under one bullet, but on second thought it doesn't add much value.
